nixos/modules/virtualisation/docker.nix

```diff
@@ -170,8 +170,9 @@ in
         "net.ipv4.conf.all.forwarding" = mkOverride 98 true;
         "net.ipv4.conf.default.forwarding" = mkOverride 98 true;
       };
-      environment.systemPackages = [ cfg.package ]
-        ++ optional cfg.enableNvidia pkgs.nvidia-docker;
+      environment.systemPackages = [ cfg.package ];
+      hardware.nvidia-container-toolkit.enable = lib.mkIf
+        (cfg.enableNvidia && cfg.package.supports-cdi) true;
```

Comment:
I think this should be `mkDefault`. Say I want to do static CDI configuration locally. Rn I'd have to `mkForce`?

Reply:
I prefer for conflicting settings to be notified to the user. As it is, if you do:

```nix
virtualisation.podman.enableNvidia = true;
hardware.nvidia-container-toolkit.enable = false;
```

You get:

```
error: The option `hardware.nvidia-container-toolkit.enable' has conflicting definition values:
- In `/nix/store/ij5zypsf2hzxa36fwxmlj1rm44cix7y7-source/hulk/hardware-configuration.nix': false
- In `/nix/store/mw5hpnrpnk1hal5n0aqsyzs952b6qkjr-source/nixos/modules/virtualisation/podman/default.nix': true
```

And I think that is good. Another option would be to use `mkDefault` and warn about those conflicting settings to notify the user, but I think this way is enough.

Reply:
Aj, so I suppose this should be instead checking for whether there's a definition for a particular "vendor" in the `virtualisation.cdi`? (because it could be static instead of dynamic)

Reply:
That would be an orthogonal issue. This change we are discussing is just so that users cannot disable `hardware.nvidia-container-toolkit` if they use `enableNvidia = true` either on docker, podman or singularity.

What you mention about static/dynamic check I think would be a nice improvement, but I would not add that complexity on the first iteration of the implementation. As the spec mentions, if that happens, the dynamic configuration will be prioritized, and everything will work fine.

Reply:
I find it confusing that `enableNvidia` must correspond to the dynamic configuration. Maybe we should rename the option? Sorry, you're clearly more aware of the context and I'm probably asking questions that should seem dumb, but I find it important to come to an understanding before we merge

Reply:
Yes, you are completely right. The naming is not optimal, I was just trying to not break user NixOS configurations. As of today, we have `virtualisation.podman.enableNvidia` and `virtualisation.docker.enableNvidia`.

This adds the runtime wrappers, but with the CDI introduction, my plan was to not break user configurations, so they don't have to change anything, and this will behind the scenes, generate the right CDI configuration for them, and they can just do `docker run --device=...` or `podman run --device=...`.

I think we can have better names, but I was prioritizing continuity, just connecting that to CDI instead of runtime wrappers, now that runtimes support CDI themselves.

Also, I don't know the deprecation rules for NixOS options if we are going down that route.

Reply:
> The naming is not optimal, I was just trying to not break user NixOS 

There's a tool for that, `mkRenamedOptionModule`. I see the point about supporting the wrappers for docker<25 though

Reply:
This sounds perfect, thank you for the pointer. I'll give it a thought and update the PR.